File: IE436_Lecture3/src/Algorithm/GraphSearch.java
package Algorithm;

import java.util.LinkedList;

public class GraphSearch {
	
	public GraphSearch() {}
	
	public void ImplementBFS(Vertex[] vertices, int s)
	{
		for(int i = 0; i < vertices.length; i++)
			vertices[i].isMarked = false;
		
		LinkedList<Integer> LIST = new LinkedList<Integer>(); 
		
		LIST.add(s);
		vertices[s].order = 1;
		vertices[s].isMarked = true;
		
		int next = 1;
		
		while(!LIST.isEmpty())
		{
			int currVertex = LIST.getFirst();
			
			boolean isAdmissible = false;
			
			for(int i = 0; i < vertices[currVertex].adj.size(); i++)
			{
				int currNeighbor = vertices[currVertex].adj.get(i);
				
				if(!vertices[currNeighbor].isMarked)
				{
					vertices[currNeighbor].isMarked = true;					
					vertices[currNeighbor].pred = currVertex;
					next++;
					vertices[currNeighbor].order = next;					
					
					LIST.addLast(currNeighbor);
					
					isAdmissible = true;
					
					break;
				}				
			}
			
			if(!isAdmissible)
				LIST.removeFirst();			
		}
		
		System.out.println();
		System.out.println("The orders in BFS");
		for(int i = 0; i < vertices.length; i++)
			System.out.println("The (order, pred) of vertex " + i + " is (" + vertices[i].order + ", " + vertices[i].pred + ")");
	}
	
	public void ImplementDFS(Vertex[] vertices, int s)
	{
		for(int i = 0; i < vertices.length; i++)
			vertices[i].isMarked = false;
		
		LinkedList<Integer> LIST = new LinkedList<Integer>(); 
		
		LIST.add(s);
		vertices[s].order = 1;
		vertices[s].isMarked = true;
		
		int next = 1;
		
		while(!LIST.isEmpty())
		{
			int currVertex = LIST.getLast();
			
			boolean isAdmissible = false;
			
			for(int i = 0; i < vertices[currVertex].adj.size(); i++)
			{
				int currNeighbor = vertices[currVertex].adj.get(i);
				
				if(!vertices[currNeighbor].isMarked)
				{
					vertices[currNeighbor].isMarked = true;					
					vertices[currNeighbor].pred = currVertex;
					next++;
					vertices[currNeighbor].order = next;					
					
					LIST.addLast(currNeighbor);
					
					isAdmissible = true;
					
					break;
				}				
			}
			
			if(!isAdmissible)
				LIST.removeLast();			
		}
		
		System.out.println();
		System.out.println("The orders in BFS");
		for(int i = 0; i < vertices.length; i++)
			System.out.println("The (order, pred) of vertex " + i + " is (" + vertices[i].order + ", " + vertices[i].pred + ")");
	}

}

File: GraphSearch.py
from Vertex import Vertex

class Search:

    def ImplementSearch(self, vertices, s, isBFS = True):

        n = len(vertices)
        for i in range(n):
            vertices[i].isMarked = False;

        LIST = []
        LIST.append(s);
        vertices[s].order = 1;
        vertices[s].isMarked = True;
        iter = 1;

        while len(LIST) != 0:

            if isBFS:
                currVertex = LIST[0]
            else:
                currVertex = LIST[-1]

            isAdmissible = False;

            for i in range(len(vertices[currVertex].adj)):
                currNeighbor = vertices[currVertex].adj[i] -1

                if not vertices[currNeighbor].isMarked:
                    vertices[currNeighbor].isMarked = True
                    vertices[currNeighbor].pred = currVertex
                    iter = iter + 1
                    vertices[currNeighbor].order = iter
                    LIST.append(currNeighbor)
                    isAdmissible = True
                    break

            if not isAdmissible:
                if isBFS:
                   del LIST[0]
                else:
                    del LIST[-1]

        print()
        if isBFS:
            print("The orders in BFS")
        else:
            print("The orders in DFS")
        for i in range(n):
            print(f"The (order, pred) of vertex {i + 1} is ({vertices[i].order}, {vertices[i].pred + 1})")
